Optimize MSW wxNotebook background painting

MaartenBent opened the pull request:
MSWDefWindowProc(WM_PAINT, ...) in OnPaint causes performance issues on large
screens so only use it when an actual custom background colour is set.

See #22308

playgithub commented:
Before and after this patch, it displays the same content, why it is much slow for large screen before this patch?

vadz commented:
Thanks again! Will merge soon.

MaartenBent commented:
I'm looking through the commit history, and this was originally added to avoid flickering in 254fbd140e67b60f46d7380c6ac242110ddd7585 in 2005.
The background colour changes where added later. I don't notice any flickering on Windows 10 without it. Maybe it was needed for XP?

vadz commented:
Yes, it seems like that the problem doesn't exist when double buffering is used. But, to be totally precise, this is not a question of (just) the Windows version, but of whether DWM is used or not, i.e. it should still be visible when using "classic theme" even under Windows 10.

But we probably could/should restrict this workaround to `!wxUxThemeIsActive()` case.

MaartenBent commented:
I simplified the patch so it only redraws when no theme is used or when a custom background is set.

vadz commented:
I actually liked the idea of only binding the event handler only if necessary (except that it would only be necessary to do it when not using themes), it seems a pity to have an event handler doing nothing 99% of the time. But we can merge it like this too, of course, if you think it's preferable.

MaartenBent commented:
I can add `wxUxThemeIsActive` to my previous solution.

vadz commented:
Please feel free to choose what you prefer -- as I said, I have a slight preference for the `Bind()`-based one, but it's really slight. Just please let me know. TIA!

MaartenBent committed:
Optimize MSW wxNotebook background painting

MSWDefWindowProc(WM_PAINT, ...) in OnPaint causes performance issues on large
screens so only use it when an actual custom background colour is set or when
no theme is used.

See #22308

MaartenBent commented:
I also prefer `Bind`, so I don't mind using that solution. I moved it outside `wxUSE_UXTHEME` in case someone would use that, please double check.
